dds/DCPS/XTypes/DynamicType.cpp

```diff
@@ -0,0 +1,119 @@
+#include "DCPS/DdsDcps_pch.h"
+#include "DynamicType.h"
+#include "MemberDescriptor.h"
+#include "TypeDescriptor.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+void DynamicType::get_descriptor(TypeDescriptor& descriptor) const
+{
+  descriptor = *descriptor_;
+}
+
+DCPS::String DynamicType::get_name() const
+{
+  return this->descriptor_->name;
+}
+
+TypeKind DynamicType::get_kind() const
+{
+  return this->descriptor_->kind;
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_name(DynamicTypeMember_rch& member, const DCPS::String& name)
+{
+  //We have a map of members in member_by_name which takes a string and gives 0 or 1 type DTMs
+  //this should use the string as the key and put the returned DTM into member.
+  DynamicTypeMembersByName::const_iterator pos = member_by_name.find(name);
+  if (pos != member_by_name.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members_by_name(DynamicTypeMembersByName& member)
+{
+  member = member_by_name;
+}
+
+DDS::ReturnCode_t DynamicType::get_member(DynamicTypeMember_rch& member, MemberId id)
+{
+  DynamicTypeMembersById::const_iterator pos = member_by_id.find(id);
+  if (pos != member_by_id.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members(DynamicTypeMembersById& member)
+{
+  member = member_by_id;
+}
+
+ACE_CDR::ULong DynamicType::get_member_count()
+{
+  return member_by_name.size();
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_index(DynamicTypeMember_rch& member, ACE_CDR::ULong index)
+{
+  if (index < member_by_index.size()) {
+    member = member_by_index[index];
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+bool DynamicType::equals(const DynamicType& other)
+{
+  return test_equality_i(*this, other);
+}
+
+bool test_equality_i(const DynamicType& lhs, const DynamicType& rhs)
+{
+//7.5.2.8.4 Operation: equals
+//Two types shall be considered equal if and only if all of their respective properties, as identified
+//in Table 54 above, are equal.
+
+//Note: We are comparing the TypeDescriptor even though the spec seems to say not to
+  DynamicTypePtrPairSeen dt_ptr_pair;
+  return test_equality(lhs, rhs, dt_ptr_pair);
+}
```

Comment:
This is another example of `*_i` functions. `test_equality_i` is currently exported, but `test_equality` is the one that does all the work. So we should swap these two function names so that `test_equality` is exported and it calls `test_equality_i`, which does the actual work.